### ISSUE TITLE
Unable to access the public APIs

### ISSUE THREAD

**Issue by rathimittharao** [cowinapi/developer.cowin#140]
How do I access the public APIs?

I get a 'Unauthorized access'

I want to access the availability of slots end point

**Comment by monish-1234** [comment]
It's mostly because you're not using headers or trying to access Protected APIs.

I'll explain the process

Here you have all the Public API Endpoints
`https://apisetu.gov.in/public/marketplace/api/cowin`

1.  Access the list of available states from 
`https://cdn-api.co-vin.in/api/v2/admin/location/states`

2. Pass the respective state id to the District API. eg for Tamil Nadu, 
`https://cdn-api.co-vin.in/api/v2/admin/location/districts/31`

3. Then send the respective District ID to whichever API Endpoint you prefer. 
Eg. To get all the appointments in Chennai for the next 7 days from today,
`https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/calendarByDistrict?district_id=571&date=10-05-2021`

4. Process the JSON according to your needs. 

TO BE NOTED : 

1. Public APIs Cache for 30 minutes. 
2. `findByPin `and `findByDistrict ` returns only un-booked slots while `calenderByPin` and `calenderByDistrict` returns both booked and un-booked slots.
3. Use respective headers for sending request else you'll get an error (refer previous issues.)
4. API has a rate limit of 100 requests for 5 mins. 
5. YOU CANNOT ACCESS PROTECTED APIs without Token. (Verify SMS OTP)

And please try refer to the previous issues before opening a new Issue.
These have been explained in detail before. I just summarized.

**Comment by rathimittharao** [comment]
Thank you @monish-1234 .. I had a typo in the useragent header. So was not working

Thanks for the detailed explanation

**Comment by sarathmekala** [comment]
@monish-1234 can you elaborate a bit more.. what headers should be sent with a request. confirmOTP sends back a token but the public api does not mention how to use it.

**Comment by monish-1234** [comment]
> @monish-1234 can you elaborate a bit more.. what headers should be sent with a request. confirmOTP sends back a token but the public api does not mention how to use it.

You don't have to confirm OTP to use public API. 
You can just access it. That's it. Those are to access protected APIs. 
And as for headers, https://github.com/cowinapi/developer.cowin/issues/34#issue-877308013 Check this out.

And be sure to hit the API from an Indian IP

**Comment by mandarsj** [comment]
i have tried from indian IP's with user agent headers. it still throws 403 forbidden. Have automated services completely blocked from accessing public api's ?

**Comment by monish-1234** [comment]
> i have tried from indian IP's with user agent headers. it still throws 403 forbidden. Have automated services completely blocked from accessing public api's ?

I don't think so . Mine is working just fine 👀

Anyway. Try this header. This is used to scrape websites.

```python
headers = {
    'authority': 'scrapeme.live',
    'dnt': '1',
    'upgrade-insecure-requests': '1',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36',
    'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'sec-fetch-site': 'none',
    'sec-fetch-mode': 'navigate',
    'sec-fetch-user': '?1',
    'sec-fetch-dest': 'document',
    'accept-language': 'en-GB,en-US;q=0.9,en;q=0.8',
}

**Comment by javatechie-Jyoti** [comment]
> i have tried from indian IP's with user agent headers. it still throws 403 forbidden. Have automated services completely blocked from accessing public api's ?

I am also facing similar issue but it seems its due to ratelimiting configured as it says "We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error". Even invoking via curl commands is giving same error (generateOTP working fine...issue coming with /states and other APIs)

**Comment by Abhi098** [comment]
> > i have tried from indian IP's with user agent headers. it still throws 403 forbidden. Have automated services completely blocked from accessing public api's ?
> 
> I am also facing similar issue but it seems its due to ratelimiting configured as it says "We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error". Even invoking via curl commands is giving same error (generateOTP working fine...issue coming with /states and other APIs)

Hi, I am also facing 403 error forbidden. Can anyone help me out how to figure this out?

**Comment by monish-1234** [comment]
> I am also facing similar issue but it seems its due to ratelimiting configured as it says "We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error". Even invoking via curl commands is giving same error (generateOTP working fine...issue coming with /states and other APIs)

This looks like a Overload on their End. Try again after sometime

**Comment by monish-1234** [comment]
> Hi, I am also facing 403 error forbidden. Can anyone help me out how to figure this out?

@Abhi098 Did you try https://github.com/cowinapi/developer.cowin/issues/140#issuecomment-836215512 this ? 

Make sure you are giving proper district code along with the date.

**Comment by Abhi098** [comment]
> > Hi, I am also facing 403 error forbidden. Can anyone help me out how to figure this out?
> 
> @Abhi098 Did you try [#140 (comment)](https://github.com/cowinapi/developer.cowin/issues/140#issuecomment-836215512) this ?
> 
> Make sure you are giving proper district code along with the date.

I was trying the simple available states API but I am facing the 403 error maybe its beacuse of the overload of requests.

**Comment by monish-1234** [comment]
@Abhi098 send me the request part of your code as it is working just fine for me

**Comment by pawanjenu** [comment]
Request made from Indian IP(droplet hosted in Bangalore location via DigitalOcean) are getting 403 forbidden, Please help

**Comment by pawanjenu** [comment]
@monish-1234 this is the curl request am making, 
`curl -L -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36' 'https://cdn-api.co-vin.in/api/v2/appointment/sessndByPin?pincode=573201&date=09-05-2021'`

**Comment by monish-1234** [comment]
@pawanjenu  Your API url is wrong. Try this

`curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`

Be sure that you're hitting the servers from an Indian IP.

**Comment by amalmurali47** [comment]
@monish-1234 You mentioned "_Public APIs Cache for 30 minutes._". 

Does that mean monitoring `/appointment/sessions/calendarByDistrict?district_id=<id>&date=<date>` is not going to work from now on? A few folks were using [the script](https://gist.github.com/amalmurali47/050ef9250c5e1cd611f2e54012d1ece3) I wrote for monitoring the appoinmtent slots.

**Comment by mandarsj** [comment]
it started working for me well, and it is still working after User-Agent header addition.Earlier probably it was an IP issue as I have deployed on azure job. now it's ok.

**Comment by monish-1234** [comment]
@amalmurali47  Most probably yes. If you setup a OTP Verification in your script (a bit of a task) like https://github.com/pallupz/covid-vaccine-booking , then you can hit the Protected API. Else you'll get unauthorized access error

**Comment by pawanjenu** [comment]
> @pawanjenu Your API url is wrong. Try this
> 
> `curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`
> 
> Be sure that you're hitting the servers from an Indian IP.

@monish-1234  i tried making an API hit to the API you gave, I get data when hitting it from local, but when trying to hit it from my cloud server(Bangalore location), I get 403 error!

please help
![image](https://user-images.githubusercontent.com/30225325/117653461-0f110200-b1b2-11eb-9d57-ab945158fcd8.png)

**Comment by javatechie-Jyoti** [comment]
I am getting the attached error for all the API calls (except generate and validate OTP). The below curl commands were working fine from my local computer till day before yesterday.

curl -X GET "https://cdn-api.co-vin.in/api/v2/admin/location/states" -H "accept: application/json" -H "Accept-Language: hi_IN"

curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/calendarByDistrict?district_id=149&date=05-05-2021" -H "accept: application/json" -H "Accept-Language: hi_IN"

![image](https://user-images.githubusercontent.com/53986175/117655495-c6a71380-b1b4-11eb-8514-7f7d78eee51d.png)

**Comment by javatechie-Jyoti** [comment]
> > I am also facing similar issue but it seems its due to ratelimiting configured as it says "We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error". Even invoking via curl commands is giving same error (generateOTP working fine...issue coming with /states and other APIs)
> 
> This looks like a Overload on their End. Try again after sometime

I am getting the attached error for all the API calls (except generate and validate OTP). The below curl commands were working fine from my local computer till day before yesterday. If its due to heavy traffic, then I am wondering how its working for you everytime :)  ..for me its giving same error at each attempt

curl -X GET "https://cdn-api.co-vin.in/api/v2/admin/location/states" -H "accept: application/json" -H "Accept-Language: hi_IN"

curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/calendarByDistrict?district_id=149&date=05-05-2021" -H "accept: application/json" -H "Accept-Language: hi_IN"

![image](https://user-images.githubusercontent.com/53986175/117655495-c6a71380-b1b4-11eb-8514-7f7d78eee51d.png)

**Comment by amalmurali47** [comment]
@monish-1234 I don't think that's true. The following request does seem to be working:

```
curl  'https://cdn-api.co-vin.in/api/v2/appointment/sessions/calendarByDistrict?district_id=17&date=10-05-2021' \
--header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'
```

**Comment by amalmurali47** [comment]
@javatechie-Jyoti You need to add a User-Agent header.

**Comment by Inderpreetsingh11** [comment]
There is a scenario on which I want to discuss. 
I subscribed to getjab.in to get notification for available slots, and I did get one that 250 slots are available in X hospital but at that very moment I ran the API for the same district but I didn't get the available slot count and after 50 seconds I was returned 0(I believe everything was booked) available slots for hospital X. 
Does anyone has any idea  about this? Why I wasn't returned that slot count at the first place?

**Comment by wittyPuneet** [comment]
> > @pawanjenu Your API url is wrong. Try this
> > `curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`
> > Be sure that you're hitting the servers from an Indian IP.
> 
> @monish-1234 i tried making an API hit to the API you gave, I get data when hitting it from local, but when trying to hit it from my cloud server(Bangalore location), I get 403 error!
> 
> please help
> ![image](https://user-images.githubusercontent.com/30225325/117653461-0f110200-b1b2-11eb-9d57-ab945158fcd8.png)

I am facing the same issue. Did you ever get it working?

**Comment by pawanjenu** [comment]
> > > @pawanjenu Your API url is wrong. Try this
> > > `curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`
> > > Be sure that you're hitting the servers from an Indian IP.
> > 
> > 
> > @monish-1234 i tried making an API hit to the API you gave, I get data when hitting it from local, but when trying to hit it from my cloud server(Bangalore location), I get 403 error!
> > please help
> > ![image](https://user-images.githubusercontent.com/30225325/117653461-0f110200-b1b2-11eb-9d57-ab945158fcd8.png)
> 
> I am facing the same issue. Did you ever get it working?

hey are you making calls from local or some cloud remote server?

**Comment by wittyPuneet** [comment]
> > > > @pawanjenu Your API url is wrong. Try this
> > > > `curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`
> > > > Be sure that you're hitting the servers from an Indian IP.
> > > 
> > > 
> > > @monish-1234 i tried making an API hit to the API you gave, I get data when hitting it from local, but when trying to hit it from my cloud server(Bangalore location), I get 403 error!
> > > please help
> > > ![image](https://user-images.githubusercontent.com/30225325/117653461-0f110200-b1b2-11eb-9d57-ab945158fcd8.png)
> > 
> > 
> > I am facing the same issue. Did you ever get it working?
> 
> hey are you making calls from local or some cloud remote server?

If I may chime in, calls from the local computer work without an issue but calls from the DO Banglore droplets fail with the above error. Did they just block DO completely?

**Comment by pawanjenu** [comment]
> > > > > @pawanjenu Your API url is wrong. Try this
> > > > > `curl -X GET "https://cdn-api.co-vin.in/api/v2/appointment/sessions/public/findByPin?pincode=573201&date=09-05-2021" -H "accept: application/json," -H "Accept-Language: hi_IN" -H 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`
> > > > > Be sure that you're hitting the servers from an Indian IP.
> > > > 
> > > > 
> > > > @monish-1234 i tried making an API hit to the API you gave, I get data when hitting it from local, but when trying to hit it from my cloud server(Bangalore location), I get 403 error!
> > > > please help
> > > > ![image](https://user-images.githubusercontent.com/30225325/117653461-0f110200-b1b2-11eb-9d57-ab945158fcd8.png)
> > > 
> > > 
> > > I am facing the same issue. Did you ever get it working?
> > 
> > 
> > hey are you making calls from local or some cloud remote server?
> 
> If I may chime in, calls from the local computer work without an issue but calls from the DO Banglore droplets fail with the above error. Did they just block DO completely?

Only a guy who's working with the co-win team can help us with that, am waiting for this to work to deploy my service,

if anyone has any idea why is this happening please let us know

**Comment by amalmurali47** [comment]
@pawanjenu @wittyPuneet COWIN APIs cannot be accessed from DigitalOcean. I believe the entire IP range has been blacklisted now. 

Results from local machine:

```
curl --location --request GET 'https://cdn-api.co-vin.in' \
--header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'
{"message":"Missing Authentication Token"}
```

This shows it's working.

Results from DigitalOcean droplet (BLR location):
```
curl --location --request GET 'https://cdn-api.co-vin.in' --header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<TITLE>ERROR: The request could not be satisfied</TITLE>
</HEAD><BODY>
<H1>403 ERROR</H1>
<H2>The request could not be satisfied.</H2>
```

Request blocked by CloudFlare.

**Comment by wittyPuneet** [comment]
> @pawanjenu @wittyPuneet COWIN APIs cannot be accessed from DigitalOcean. I believe the entire IP range has been blacklisted now.
> 
> Results from local machine:
> 
> ```
> curl --location --request GET 'https://cdn-api.co-vin.in' \
> --header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'
> {"message":"Missing Authentication Token"}
> ```
> 
> This shows it's working.
> 
> Results from DigitalOcean droplet (BLR location):
> 
> ```
> curl --location --request GET 'https://cdn-api.co-vin.in' --header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36'
> 
> <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
> <HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
> <TITLE>ERROR: The request could not be satisfied</TITLE>
> </HEAD><BODY>
> <H1>403 ERROR</H1>
> <H2>The request could not be satisfied.</H2>
> ```
> 
> Request blocked by CloudFlare.

Are you aware of other providers that work? AWS, Linode, etc?

**Comment by amalmurali47** [comment]
@wittyPuneet Yes. AWS, Linode both work. I just tested them.

**Comment by arhamshah** [comment]
I recently faced the same issue of response as 403: Forbidden.
I solved it by adding a request header of user-agent.
`headers= {'user-agent': 'Mozilla/5.0 (Windows; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36'}`
